src/category_theory/triangulated/pretriangulated.lean

```diff
@@ -0,0 +1,206 @@
+/-
+Copyright (c) 2021 Luke Kershaw. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Luke Kershaw
+-/
+import category_theory.additive.basic
+import category_theory.shift
+import category_theory.preadditive.additive_functor
+import category_theory.triangulated.basic
+import category_theory.triangulated.rotate
+
+/-!
+# Pretriangulated Categories
+
+This file contains the definition of pretriangulated categories and triangulated functors
+between them.
+
+TODO: generalise this to n-angulated categories as in https://arxiv.org/abs/1006.4592
+-/
+
+noncomputable theory
+
+open category_theory
+open category_theory.preadditive
+
+universes v v₀ v₁ v₂ u u₀ u₁ u₂
+
+namespace category_theory.triangulated
+open category_theory.category
+
+/-
+We work in an additive category `C` equipped with an additive shift.
+-/
+variables (C : Type u) [category.{v} C] [has_shift C] [additive_category C]
+[functor.additive (shift C).functor] [functor.additive (shift C).inverse]
+
+/--
+An additive category `C` with an additive shift, and a class of "distinguished triangles"
+relative to that shift is called pretriangulated if the following hold:
+* Any triangle that is isomorphic to a distinguished triangle is also distinguished.
+* Any triangle of the form `(X,X,0,id,0,0)` is distinguished.
+* For any morphism `f: X ⟶ Y` there exists a distinguished triangle of the form `(X,Y,Z,f,g,h)`.
+* The triangle `(X,Y,Z,f,g,h)` is distinguished if and only if `(Y,Z,X⟦1⟧,g,h,-f⟦1⟧)` is.
+* Given a commutative diagram:
+```
+        f       g       h
+    X  ───> Y  ───> Z  ───> X⟦1⟧
+    │       │                │
+    │a      │b               │a⟦1⟧'
+    V       V                V
+    X' ───> Y' ───> Z' ───> X'⟦1⟧
+        f'      g'      h'
+```
+  whose rows are distinguished triangles, there exists a morphism `c: Z ⟶ Z'` such that `(a,b,c)`
+  is a triangle morphism.
+See https://stacks.math.columbia.edu/tag/0145
+-/
+class pretriangulated :=
+(distinguished_triangles : set(triangle C))
+(isomorphic_distinguished : Π (T₁ ∈ distinguished_triangles) (T₂ : triangle C) (T₁ ≅ T₂),
+  T₂ ∈ distinguished_triangles)
+(contractible_distinguished : Π (X : C), (contractible_triangle C X) ∈ distinguished_triangles)
+(distinguished_cocone_triangle : Π (X:C) (Y:C) (f: X ⟶ Y), (∃ (Z : C) (g : Y ⟶ Z) (h : Z ⟶ X⟦1⟧),
+  ({ obj₁ := X, obj₂ := Y, obj₃ := Z, mor₁ := f, mor₂ := g, mor₃ := h} : triangle C) ∈
```

Comment:
It would be nice to be able to write `triangle.mk f g h` instead of `({ obj₁ := X, obj₂ := Y, obj₃ := Z, mor₁ := f, mor₂ := g, mor₃ := h} : triangle C)`. 

(Presumably this would be useful earlier, as well.)